neuralpredictors/training/tracking.py

```diff
@@ -146,3 +153,360 @@ def asdict(self, time_key="time", make_copy=True):
+
+        """
+        self.main_objective = tracker_dict["main_objective"]
+        self.objectives = tracker_dict["objectives"]
+        self.log = tracker_dict["log"]
+        self.time = tracker_dict["time"]
+        self.epoch = tracker_dict["epoch"]
+
+    @classmethod
+    def from_dict(cls, tracker_dict: Dict) -> "AdvancedTracker":
+        """
+        Same as `load_state_dict`, but creates a new tracker from scratch.
+        Args:
+            tracker_dict: state of tracker that should be loaded in new tracker
+
+        Returns:
+            new tracker with `tracker_dict` loaded
+
+        """
+        tracker = cls(main_objective=tracker_dict["main_objective"], **tracker_dict["objectives"])
+        tracker.log = tracker_dict["log"]
+        tracker.time = tracker_dict["time"]
+        tracker.epoch = tracker_dict["epoch"]
+        return tracker
+
+    def _initialize_log(self, objectives) -> Dict:
+        log = {}
+        for key, objective in objectives.items():
+            if isinstance(objective, dict):
+                log[key] = self._initialize_log(objective)
+            elif not callable(objective):
+                log[key] = []
+        return log
+
+    def _initialize_epoch(self, log, objectives):
+        """
+        For each key in `objectives`, go through log and append a new entry to its list.
+        The entry reflects the default value saved in `objectives`.
+        Args:
+            log: sub dictionary to add the objectives to
+            objectives: dictionary of default values
+        """
+        for key, objective in objectives.items():
+            if isinstance(objective, dict):
+                self._initialize_epoch(log[key], objective)
+            elif not callable(objective):
+                while len(log[key]) <= self.epoch:
+                    log[key].append(objective)
+
+    def _log_objective_value(self, value: float, log: Dict, key: Tuple[str, ...] = ()):
+        """
+        Recursively walk through the log dictionary to get to follow `key`.
+        When on lowest level: add `value` to entry at current epoch.
+        Args:
+            value: objective value to log
+            log: log subdict to add value to
+            key: key for where `value` is saved
+        """
+        if len(key) > 1:
+            self._log_objective_value(value, log[key[0]], key[1:])
+        else:
+            log[key[0]][self.epoch] += value
+
+    def _log_objective_callables(self, log: Dict, objectives: Dict):
+        """
+        Log objectives that are specified as callables
+        Disclaimer: this is not very well tested!
+        Args:
+            log:
+            objectives:
+        """
+        for key, objective in objectives.items():
+            if isinstance(objective, dict):
+                self._log_objective_callables(log[key], objective)
+            elif callable(objective):
+                log[key].append(objective())
```

Comment:
I assume a callable objective is supposed to return the objective value, correct? Why is this value appended here instead of being added?

I would have expected something like `log[key][self.epoch] += objective()` here.